projects/packages/image-cdn/src/class-image-cdn-image-sizes.php

```diff
@@ -0,0 +1,191 @@
+ * The Image Sizes library.
+ *
+ * @package automattic/jetpack-image-cdn
+ */
+
+namespace Automattic\Jetpack\Image_CDN;
+
+/**
+ * Class Image_CDN_Image_Sizes
+ *
+ * Manages image resizing via Jetpack CDN Service.
+ */
+class Image_CDN_Image_Sizes {
+
+	/**
+	 * Attachment metadata.
+	 *
+	 * @var array
+	 */
+	public $data;
+
+	/**
+	 * Image to be resized.
+	 *
+	 * @var Image
+	 */
+	public $image;
+
+	/**
+	 * Intermediate sizes.
+	 *
+	 * @var null|array
+	 */
+	public static $sizes = null;
+
+	/**
+	 * Construct new sizes meta
+	 *
+	 * @param int   $attachment_id Attachment ID.
+	 * @param array $data          Attachment metadata.
+	 */
+	public function __construct( $attachment_id, $data ) {
+		$this->data  = $data;
+		$this->image = new Image_CDN_Image( $data, get_post_mime_type( $attachment_id ) );
+		$this->generate_sizes();
+	}
+
+	/**
+	 * Generate sizes for attachment.
+	 *
+	 * @return array Array of sizes; empty array as failure fallback.
+	 */
+	protected function generate_sizes() {
+
+		// There is no need to generate the sizes a new for every single image.
+		if ( null !== self::$sizes ) {
+			return self::$sizes;
+		}
+
+		/*
+		 * The following logic is copied over from wp_generate_attachment_metadata
+		 */
+		$_wp_additional_image_sizes = wp_get_additional_image_sizes();
+
+		$sizes = array();
+
+		$intermediate_image_sizes = get_intermediate_image_sizes();
+
+		foreach ( $intermediate_image_sizes as $s ) {
+			$sizes[ $s ] = array(
+				'width'  => '',
+				'height' => '',
+				'crop'   => false,
+			);
+			if ( isset( $_wp_additional_image_sizes[ $s ]['width'] ) ) {
+				// For theme-added sizes.
+				$sizes[ $s ]['width'] = (int) $_wp_additional_image_sizes[ $s ]['width'];
+			} else {
+				// For default sizes set in options.
+				$sizes[ $s ]['width'] = get_option( "{$s}_size_w" );
+			}
+
+			if ( isset( $_wp_additional_image_sizes[ $s ]['height'] ) ) {
+				// For theme-added sizes.
+				$sizes[ $s ]['height'] = (int) $_wp_additional_image_sizes[ $s ]['height'];
+			} else {
+				// For default sizes set in options.
+				$sizes[ $s ]['height'] = get_option( "{$s}_size_h" );
+			}
+
+			if ( isset( $_wp_additional_image_sizes[ $s ]['crop'] ) ) {
+				// For theme-added sizes.
+				$sizes[ $s ]['crop'] = $_wp_additional_image_sizes[ $s ]['crop'];
+			} else {
+				// For default sizes set in options.
+				$sizes[ $s ]['crop'] = get_option( "{$s}_crop" );
+			}
+		}
+
+		self::$sizes = $sizes;
+
+		return $sizes;
+	}
+
+	/**
+	 * Add filtered sizes.
+	 *
+	 * @return array
+	 */
+	public function filtered_sizes() {
+		// Remove filter preventing the creation of advanced sizes.
+		remove_filter(
+			'intermediate_image_sizes_advanced',
+			array( 'Jetpack_Photon', 'filter_photon_noresize_intermediate_sizes' )
+		);
+
+		/** This filter is documented in wp-admin/includes/image.php */
+		$sizes = apply_filters( 'intermediate_image_sizes_advanced', self::$sizes, $this->data );
+
+		// Re-add the filter removed above.
```

Comment:
Looks like that the intention is that this also depends on the code existing to disable Jetpack's copy of Photon when the package is used?

Reply:
Yes, this relies on the inbuilt Jetpack-Photon being deactivated by Boost's compatibility code to prevent its `intermediate_image_sizes_advanced` hooks from getting in the way.

When `jetpack_photon_noresize_mode` is enabled (via a filter) and I hack out Boost's code to deactivate Jetpack Photon, this breaks.

Just to be extra-defensive I have committed e5d7fd1728 which removes the inbuilt-Jetpack version of this filter just-in-case.